src/include/libpmemobj.h

```diff
@@ -84,6 +99,505 @@ void pmemobj_set_funcs(
 		void *(*malloc_func)(size_t size),
 		void (*free_func)(void *ptr));
 
+/*
+ * Locking...
+ */
+#define	PMEMOBJ_CL_ALIGNMENT 64 /* cache line alignment for performance */
+
+typedef union padded_pmemmutex {
+	char padding[PMEMOBJ_CL_ALIGNMENT];
+	struct {
+		uint64_t runid;
+		pthread_mutex_t mutex;
+	} pmemmutex;
+} PMEMmutex;
+
+typedef union padded_pmemrwlock {
+	char padding[PMEMOBJ_CL_ALIGNMENT];
+	struct {
+		uint64_t runid;
+		pthread_rwlock_t rwlock;
+	} pmemrwlock;
+} PMEMrwlock;
+
+typedef union padded_pmemcond {
+	char padding[PMEMOBJ_CL_ALIGNMENT];
+	struct {
+		uint64_t runid;
+		pthread_cond_t cond;
+	} pmemcond;
+} PMEMcond;
+
+int pmemobj_mutex_zero(PMEMmutex *mutexp);
+int pmemobj_mutex_lock(PMEMobjpool *pop, PMEMmutex *mutexp);
+int pmemobj_mutex_trylock(PMEMobjpool *pop, PMEMmutex *mutexp);
+int pmemobj_mutex_unlock(PMEMobjpool *pop, PMEMmutex *mutexp);
+
+int pmemobj_rwlock_zero(PMEMrwlock *rwlockp);
+int pmemobj_rwlock_rdlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_wrlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_timedrdlock(PMEMobjpool *pop,
+	PMEMrwlock *restrict rwlockp,
+	const struct timespec *restrict abs_timeout);
+int pmemobj_rwlock_timedwrlock(PMEMobjpool *pop,
+	PMEMrwlock *restrict rwlockp,
+	const struct timespec *restrict abs_timeout);
+int pmemobj_rwlock_tryrdlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_trywrlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+int pmemobj_rwlock_unlock(PMEMobjpool *pop, PMEMrwlock *rwlockp);
+
+int pmemobj_cond_zero(PMEMcond *condp);
+int pmemobj_cond_broadcast(PMEMobjpool *pop, PMEMcond *condp);
+int pmemobj_cond_signal(PMEMobjpool *pop, PMEMcond *condp);
+int pmemobj_cond_timedwait(PMEMobjpool *pop, PMEMcond *restrict condp,
+	PMEMmutex *restrict mutexp, const struct timespec *restrict abstime);
+int pmemobj_cond_wait(PMEMobjpool *pop, PMEMcond *condp,
+	PMEMmutex *restrict mutexp);
+
+/* XXX static inline */
```

Comment:
By making it inline, we could possibly improve performance a bit, but this also requires exposing some implementation details in the public header.